staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -38,10 +39,109 @@ import (
 	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
 	"k8s.io/apiserver/pkg/endpoints/metrics"
 	endpointsrequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storage"
 
-	klog "k8s.io/klog/v2"
+	"k8s.io/klog/v2"
 )
 
+// watchListEncoder performs encoding of
+// a special watchList bookmark event
+//
+// see: https://github.com/kubernetes/kubernetes/pull/127319
+type watchListEncoder struct {
+	initialEventsListBlueprint runtime.Object
+	targetGVK                  *schema.GroupVersionKind
+
+	rawEncoder    runtime.Encoder
+	objectEncoder runtime.Encoder
+
+	buffer     runtime.Splice
+	identifier runtime.Identifier
+}
+
+func newWatchListEncoder(initialEventsListBlueprint runtime.Object,
+	targetGVK *schema.GroupVersionKind,
+	rawEncoder runtime.Encoder,
+	objectEncoder runtime.Encoder) *watchListEncoder {
+	return &watchListEncoder{
+		initialEventsListBlueprint: initialEventsListBlueprint,
+		targetGVK:                  targetGVK,
+		rawEncoder:                 rawEncoder,
+		objectEncoder:              objectEncoder,
+		buffer:                     runtime.NewSpliceBuffer(),
+	}
+}
+
+func (e *watchListEncoder) Encode(obj runtime.Object, w io.Writer) error {
+	// Note that we don't check if the object supports caching its serializations
```

Comment:
This comment isn't fully true.

This encoder is actually getting every single object that is being send via a watch request. It's just we simply delegate serialization of any non-bookmark objects to the underlying serializer.
And we rely on the fact that objects for bookmarks are not wrapper into CachingObject...